[Unified PDF] Data Detectors are offset from detected items on embedded PDFs

pxlcoder opened the pull request:
#### bbd3a8dd40bdb2cbbe6d8f20cf98bf18ee430619
<pre>
[Unified PDF] Data Detectors are offset from detected items on embedded PDFs
<a href="https://bugs.webkit.org/show_bug.cgi?id=272684">https://bugs.webkit.org/show_bug.cgi?id=272684</a>
<a href="https://rdar.apple.com/126232906">rdar://126232906</a>

Reviewed by Abrar Rahman Protyasha.

276131@main added support for Data Detectors for Unified PDF. However, the
geometry used for highlights was using the plugin&apos;s coordinate space. This is
incorrect for embedded PDFs, as the `PageOverlay` uses `OverlayType::Document`.

Fix by ensuring all highlight geometry uses the main frame&apos;s contents
coordinate space.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::lastKnownMousePositionInView const):
(WebKit::PDFPluginBase::mousePositionInView const): Deleted.

Revert the change made in 276131@main, as this method now has a single use,
inside the same class.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
(WebKit::PDFDataDetectorOverlayController::createPlatformDataDetectorHighlight const):
(WebKit::PDFDataDetectorOverlayController::handleMouseEvent):
(WebKit::PDFDataDetectorOverlayController::handleDataDetectorAction):

Click information should be in the window&apos;s coordinate space.

(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::mainFrameView const):

Expose the main frame&apos;s `FrameView`, so that the overlay controller can get
the `visibleContentRect`, and convert between coordinate spaces.

(WebKit::UnifiedPDFPlugin::rectForSelectionInMainFrameContentsSpace const):

Similar to `ServicesOverlayController`, the only way to convert to document
coordinates is to call `contentsToWindow` and then `windowToContents`.

(WebKit::UnifiedPDFPlugin::rectForSelectionInPluginSpace const): Deleted.

Canonical link: <a href="https://commits.webkit.org/277510@main">https://commits.webkit.org/277510@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f26473250baf6b76e623059efcb5a1ea818a6d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26992 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50481 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43853 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32860 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24477 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38908 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48380 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24672 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41307 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-api-tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20200 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22150 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5847 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44163 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42920 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52375 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22835 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19193 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46209 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 41 flakes 121 failures; Uploaded test results; 9 flakes 72 failures; Running compile-webkit-without-change") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24107 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45250 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10549 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24897 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23828 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
```diff
@@ -128,20 +132,25 @@
     updateDataDetectorHighlightsIfNeeded(pageIndex);
 
-    auto mousePositionInPluginSpace = plugin->mousePositionInView(event);
+    auto mousePositionInWindowSpace = event.position();
```
nit: What do you think about renaming this to `mousePositionInRootView`, since that's the nomenclature we've been rolling with in UnifiedPDF?

aprotyas replied:
(repeat for other `mousePositionInWindowSpace` variables in this PR)

pxlcoder replied:
I think that might be a little strange, since the variable is passed into `windowToContents`.

aprotyas replied:
Hmm, yes, `mousePositionInWindowSpace` reads better given how we're using it. It might seem like "yet another coordinate space" for an uninitiated reader, although I don't know how valid that claim is.

aprotyas replied:
(Feel free to ignore!)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3002,17 +3013,22 @@ static FloatRect computeMarqueeSelectionRect(const WebCore::FloatPoint& point1,
-    return convertUp(CoordinateSpace::PDFPage, CoordinateSpace::Plugin, FloatRect { [selection boundsForPage:page.get()] }, *pageIndex);
+    auto rectForSelectionInRootView = this->rectForSelectionInRootView(selection);
+    auto rectForSelectionInContents = localFrameView->rootViewToContents(rectForSelectionInRootView);
+    return mainFrameView->windowToContents(localFrameView->contentsToWindow(roundedIntRect(rectForSelectionInContents)));
```
At first glance, it's confusing that `windowToContents(contentsToWindow(X)) =/= X`!

pxlcoder committed:
[Unified PDF] Data Detectors are offset from detected items on embedded PDFs
https://bugs.webkit.org/show_bug.cgi?id=272684
rdar://126232906

Reviewed by Abrar Rahman Protyasha.

276131@main added support for Data Detectors for Unified PDF. However, the
geometry used for highlights was using the plugin's coordinate space. This is
incorrect for embedded PDFs, as the `PageOverlay` uses `OverlayType::Document`.

Fix by ensuring all highlight geometry uses the main frame's contents
coordinate space.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::lastKnownMousePositionInView const):
(WebKit::PDFPluginBase::mousePositionInView const): Deleted.

Revert the change made in 276131@main, as this method now has a single use,
inside the same class.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
(WebKit::PDFDataDetectorOverlayController::createPlatformDataDetectorHighlight const):
(WebKit::PDFDataDetectorOverlayController::handleMouseEvent):
(WebKit::PDFDataDetectorOverlayController::handleDataDetectorAction):

Click information should be in the window's coordinate space.

(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::mainFrameView const):

Expose the main frame's `FrameView`, so that the overlay controller can get
the `visibleContentRect`, and convert between coordinate spaces.

(WebKit::UnifiedPDFPlugin::rectForSelectionInMainFrameContentsSpace const):

Similar to `ServicesOverlayController`, the only way to convert to document
coordinates is to call `contentsToWindow` and then `windowToContents`.

(WebKit::UnifiedPDFPlugin::rectForSelectionInPluginSpace const): Deleted.

Canonical link: https://commits.webkit.org/277510@main
